### DISCUSSION TITLE
Issues using async transform to fetch data and render Tag

### DISCUSSION THREAD

**Discussion by cryptobench** [markdoc/markdoc#443] [Q&A] [answered]
Heya.

I am trying to create a tag that will fetch a code example from Github, and I think i'm almost there. I have an issue where if I defined this `github` Tag below, then my pages fails to render due to the fact that `nodes` inside `pageProps.markdoc.content` suddenly returns an empty `{}` instead of an array with content. If I remove the Tag then everything is fine again. 

Error from my app.js after including the Tag in the code:
```
TypeError: nodes is not iterable
  27 | let sections = []
  28 | console.log('nodes', nodes)
> 29 | for (let node of nodes) {
     |                 ^
  30 |   if (['h1', 'h2', 'h3', 'h4', 'h5', 'h6'].includes(node.name)) {
  31 |     let title = getNodeText(node)
  32 |     if (title) {

```

Do you know what i'm doing wrong in my transform code? 

```js
  github: {
    render: GithubCode,
    selfClosing: true,
    attributes: {
      githubPath: { type: String },
    },
    async transform(node, config) {
      const { githubPath } = node.attributes
      if (!githubPath) return null
      try {
        const code = await fetch(
          `https://raw.githubusercontent.com/someUser/someRepo/master/${githubPath}`
        )
        const src = await code.text()
        const attributes = node.transformAttributes(config)
        return new Tag(
          this.render,
          { ...attributes, code },
          node.transformChildren(config)
        )
      } catch (e) {
        console.error(e)
        return null
      }
    },
  },
 ```
 
 ```js
export function GithubCode({ children, code, language = 'js' }) {
  return <p>{code}</p>
}

```

**Comment by rpaul-stripe** [comment]
There is a small error in your transform function — it passes the resolved `code` promise into the tag when I assume you want the `src` variable. But I'm not sure if that's actually causing the error that you are seeing. When I put a minimal version of the tag into a [standalone example](https://gist.github.com/rpaul-stripe/e15f174034f9217b5b2b3b517e294bb3), it seems to work as intended. There might be something else going on in your `app.js` that is causing the problem.

**Reply by cryptobench** [reply]
It does sound like it's in the app.js then.

```js
import Head from 'next/head'
import { slugifyWithCounter } from '@sindresorhus/slugify'
import { Inter } from 'next/font/google'
const inter = Inter({ subsets: ['latin'] })
import { GoogleAnalytics } from 'nextjs-google-analytics'

import { Layout } from '@/components/Layout'

import 'focus-visible'
import '@/styles/tailwind.css'
function getNodeText(node) {
  let text = ''
  for (let child of node.children ?? []) {
    if (typeof child === 'string') {
      text += child
    }
    text += getNodeText(child)
  }
  return text
}

function collectHeadings(
  nodes,
  slugify = slugifyWithCounter(),
  lastNodes = []
) {
  let sections = []
  for (let node of nodes) {
    if (node.name === 'Heading') {
      let { level, id } = node.attributes
      let title = getNodeText(node)
      if (title) {
        let newNode = { ...node.attributes, title, children: [] }
        if (lastNodes[level - 2]) {
          lastNodes[level - 2].children.push(newNode)
        } else {
          sections.push(newNode)
        }
        lastNodes[level - 1] = newNode
        lastNodes.length = level
      }
    }
    sections.push(...collectHeadings(node.children ?? [], lastNodes))
  }
  return sections
}

export default function App({ Component, pageProps }) {
  let title = pageProps.markdoc?.frontmatter.title
  let type = pageProps.markdoc?.frontmatter.type
  let tags = pageProps.markdoc?.frontmatter.tags

  let pageTitle =
    pageProps.markdoc?.frontmatter.pageTitle ||
    `${pageProps.markdoc?.frontmatter.title}`

  let description = pageProps.markdoc?.frontmatter.description
  let tableOfContents = pageProps.markdoc?.content
    ? collectHeadings(pageProps.markdoc.content)
    : []

  return (
    <>
      <GoogleAnalytics trackPageViews />

      <div className={inter.className}>
        <Head>
          
          <title>{pageTitle}</title>
          {description && <meta name="description" content={description} />}
        </Head>
        <Layout
          title={title}
          tableOfContents={tableOfContents}
          type={type}
          tags={tags}
        >
          <Component {...pageProps} />
        </Layout>
      </div>
    </>
  )
}

```


Its specifically when I try to iterate over `nodes` that it fails. I tried console.logging it and now its just an empty `{}` instead of an array with items.

**Reply by cryptobench** [reply]
I just tried logging the markdoc content and it's completely empty after adding the async transform function

`console.log('content', pageProps.markdoc.content)` I just tried logging the markdoc content and it's completely empty after adding the async transform function

`console.log('content', pageProps.markdoc.content)`

**Reply by mfix-stripe** [reply]
@cryptobench would it be possible for you to upload your full code somewhere, or a reproducible example, and I can take a look at fixing it?

**Reply by cryptobench** [reply, marked as answer]
@mfix-stripe I think I just got it fixed right now. I was using the TailwindUI template which had Markdoc version 1.6.7, so veryyyy old. I just upgraded it and updated my app.js, so that it now uses a try/catch block to try and find nodes with h1, h2, h3, h4, h5, h6 tags and if not then just render an empty array. This solved my issue.

**Reply by mfix-stripe** [reply]
Great! Yeah older versions of `@markdoc/next.js` didn't support async transforms.

**Reply by cryptobench** [reply]
Nope, I remember reading about the experimental async support in version 1.something , but it didn't strike my any thought what version I was using in my project. Fast forward a few hours and I got the thought about it, and it solved it. 

@mfix-stripe I really appreciate the swift responses you are providing. Thank you so much!

**Reply by mfix-stripe** [reply]
Glad you fixed it @cryptobench! We are still up in the air about whether we want to document this as a non-experimental feature. I'll keep you posted.
